### ISSUE TITLE
Transitive imports in extraPaths are not resolved

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#1522]
Note: this might be a duplicate of #1515, but since this happens within one workspace and also with absolute paths it looks different enough to open a separate issue.

**Describe the bug**
Transitive imports in `extraPaths` are not resolved.

**To Reproduce**
Create the following file structure:

```
.
├── pyrightconfig.json
├── src1
│   └── foo.py
└── src2
    └── bar
        ├── baz.py
        └── __init__.py
```

`pyrightconfig.json`:
```json
{
    "executionEnvironments": [
        {
            "root": "src1",
            "extraPaths": ["src2"]
        }
    ]
}
```

`src1/foo.py`
```python
import bar
```

`src2/bar/__init__.py`
```python
import bar.baz
```

**Expected behavior**
No import error.

**Screenshots or Code**

```
$ pyright --verbose src1
Loading configuration file at /tmp/pyright4/root/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
Search paths found for configured python interpreter:
  /tmp/python-3.8/lib/python3.8
  /tmp/python-3.8/lib/python3.8/lib-dynload
  /tmp/python-3.8/lib/python3.8/site-packages
When attempting to get search paths from python interpreter:
  Executing interpreter: 'python3'
  Skipping '/tmp/python-3.8/lib/python38.zip' because it is not a valid directory
  Received 3 paths from interpreter
    /tmp/python-3.8/lib/python3.8
    /tmp/python-3.8/lib/python3.8/lib-dynload
    /tmp/python-3.8/lib/python3.8/site-packages
stubPath /tmp/pyright4/root/typings is not a valid directory.
Searching for source files
Found 1 source file
Could not import 'bar.baz' in file '/tmp/pyright4/root/src2/bar/__init__.py'
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/tmp/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking in stubPath '/tmp/pyright4/root/typings'
  Attempting to resolve using root path '/tmp/pyright4/root/typings'
  Looking in root directory of execution environment '/tmp/pyright4/root'
  Attempting to resolve using root path '/tmp/pyright4/root'
  Looking in python search path '/tmp/python-3.8/lib/python3.8'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8'
  Looking in python search path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Looking in python search path '/tmp/python-3.8/lib/python3.8/site-packages'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/site-packages'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/site-packages'
  Looking for typeshed path
  Looking for typeshed stubs path
  Typeshed path not found
0 errors, 0 warnings, 0 infos
Completed in 0.515sec
```

**VS Code extension or command-line**

```
$ pyright --version
pyright 1.1.109
```

**Additional context**
Running `PYTHONPATH=src2 python src1/foo.py` works fine, without import error.

**Comment by jakebailey** [comment]
Extra paths are rooted relative to the root of the execution environment (or VS Code workspace), not relative to the config file. Try doing `../src2`.

**Comment by erictraut** [comment]
No, extra paths are relative to the config file, not the execution environment.

**Comment by erictraut** [comment]
Here's the code that implements it: https://github.com/microsoft/pyright/blob/d9b21f0d690ac0629ddf64b71ba8609cfe5ddc1b/packages/pyright-internal/src/common/configOptions.ts#L1440
You can see that it's using the project path.

**Comment by jakebailey** [comment]
Hmm. I guess I'll have to debug #1515 and this, then. They feel the same to me, and it's strange that the verbose logs aren't listing _any_ folder at all for that added extraPath.

**Comment by erictraut** [comment]
@ipkiss42, I tried your repro steps. It is working as intended. You have defined one execution environment root (`src1`). All files under that root will be analyzed using the rules for that execution environment. All source files that are not under that root but still within the workspace will not be analyzed with those rules. Since `baz.py` is not under the `src1` root, it does not get any special extra import paths, and all imports are resolved from the root directory of the workspace.

If you want `src1` and `src2` to both be analyzed as separate roots, then you need to update your pyrightconfig.json as follows:
```
{
  "executionEnvironments": [
    {
      "root": "src1",
      "extraPaths": ["src2"]
    },
    {
      "root": "src2"
    }
  ]
}
```

It would also work to specify a single execution environment with `src2` as a extraPath.
```
{
  "executionEnvironments": [
    {
      "root": ".",
      "extraPaths": ["src2"]
    }
  ]
}
```

**Comment by erictraut** [comment]
I'm going to close this issue. If you think that I have misdiagnosed the problem or have additional questions, feel free to post.

**Comment by ipkiss42** [comment]
I'm confused by what "analyze" means in this context.

I make a distinction (maybe wrongly?) between "parsed and analyzed, with errors being reported" (which is what happens with an execution environment root, IIUC) and "parsed because it is a transitive dependency of an analyzed module" (which I expected was a pre-requisite for proper analysis of the analyzed module).

My intent was to analyse (i.e. report errors on) the files in `src1` only, but with all the transitive imports still being taken into account, to avoid spurious type errors.

So from my understanding:
* With the configuration I posted initially, transitive imports will not be resolved because `src2` is not a root of en execution environment. I can thus get false positives in the pyright output.
* With both of your configurations, pyright will also check deeply `src2`. I can thus get errors reported for `src2`, which I don't really care about (at least for now).

Am I missing something?
Why wouldn't transitive imports be looked for in the `extraPaths`?

**Comment by ipkiss42** [comment]
@erictraut: I'm not sure if you still monitor closed issues, so I thought I would ping you directly. Sorry if this ends up being redundant!

**Comment by erictraut** [comment]
Imports within a source file are resolved according to the import paths defined by the execution environment to which that file belongs. A source file belongs to (at most) one execution environment — the first one whose root directory contains the file.

**Comment by ipkiss42** [comment]
I understand this is the current behavior, but I'm not sure if this is desirable. Why wouldn't transitive imports be resolved according to the **current** execution environment (i.e. the one for the initial file being analyzed)?

It looks like another side-effect of #1508 (the inability to analyze the same file with different venvs), . Would you say that this limitation could be lifted, if #1508 was solved in favor of allowing files to be parsed by more than one execution environment?

**Comment by erictraut** [comment]
I understand that you want this feature to behave in a different way than it was designed.

Yes, this is part of the discussion in #1508. 

As I mentioned in #1508, we already have a separate mechanism in place for providing the behavior you're looking for. It involves instantiating the type checker once for each environment that you would like to analyze. This involves separate pyrightconfig.json files for each such project and a multi-root workspace if you're using VS Code.

**Comment by ipkiss42** [comment]
Thank you. I'll follow #1508 and see where it lands.
